pp/types/peer_list.go

```diff
@@ -0,0 +1,382 @@
+	Status             int
+}
+
+func (peerInfo *PeerInfo) String() string {
+	return fmt.Sprintf("{%v(%v)}", peerInfo.P2pAddress, peerInfo.NetworkAddress)
+}
+
+type PeerList struct {
+	localNetworkAddress   string
+	ppListPath            string
+	ppMapByNetworkAddress *sync.Map // map[string]*PeerInfo
+	ppMapByP2pAddress     *sync.Map // map[string]*PeerInfo
+
+	rwmutex sync.RWMutex
+}
+
+func (peerList *PeerList) Init(localNetworkAddress, ppListPath string) {
+	peerList.localNetworkAddress = localNetworkAddress
+	peerList.ppListPath = ppListPath
+	peerList.ppMapByNetworkAddress = &sync.Map{}
+	peerList.ppMapByP2pAddress = &sync.Map{}
+}
+
+func (peerList *PeerList) loadPPListFromFile() error {
+	csvFile, err := os.OpenFile(peerList.ppListPath, os.O_CREATE|os.O_RDWR, 0777)
+	defer csvFile.Close()
+	if err != nil {
+		return errors.Wrap(err, "LoadPPListFromFile cannot open ppList file")
+	}
+
+	reader := csv.NewReader(csvFile)
+	reader.FieldsPerRecord = -1
+	record, err := reader.ReadAll()
+	if err != nil {
+		return errors.Wrap(err, "LoadPPListFromFile cannot decode csv file")
+	}
+
+	for _, item := range record {
+		if len(item) < 5 {
+			utils.ErrorLogf("LoadPPListFromFile ppList record is incomplete. %v fields (%v expected)", len(item), 5)
+			continue
+		}
+		networkID, err := types.IDFromString(item[0])
+		if err != nil {
+			utils.ErrorLog("LoadPPListFromFile invalid networkId ["+item[0]+"]", err)
+			continue
+		}
+
+		discoveryTime, err := strconv.ParseInt(item[3], 10, 64)
+		if err != nil {
+			utils.ErrorLog("LoadPPListFromFile invalid discoveryTime ["+item[3]+"]", err)
+			continue
+		}
+
+		lastConnectionTime, err := strconv.ParseInt(item[4], 10, 64)
+		if err != nil {
+			utils.ErrorLog("LoadPPListFromFile invalid lastConnectionTime ["+item[4]+"]", err)
+			continue
+		}
+
+		pp := &PeerInfo{
+			NetworkAddress:     networkID.NetworkAddress,
+			P2pAddress:         networkID.P2pAddress,
+			RestAddress:        item[1],
+			WalletAddress:      item[2],
+			DiscoveryTime:      discoveryTime,
+			LastConnectionTime: lastConnectionTime,
+			Status:             PEER_NOT_CONNECTED,
+		}
+
+		peerList.ppMapByNetworkAddress.Store(pp.NetworkAddress, pp)
+		peerList.ppMapByP2pAddress.Store(pp.P2pAddress, pp)
+	}
+	return nil
+}
+
+func (peerList *PeerList) savePPListToFile() error {
+	peerList.rwmutex.Lock()
+	defer peerList.rwmutex.Unlock()
+
+	err := os.Truncate(peerList.ppListPath, 0)
+	if err != nil {
+		return err
+	}
+	csvFile, err := os.OpenFile(peerList.ppListPath, os.O_CREATE|os.O_RDWR, 0777)
+	if err != nil {
+		return err
+	}
+	defer csvFile.Close()
+	writer := csv.NewWriter(csvFile)
+
+	linesWritten := 0
+	peerList.ppMapByNetworkAddress.Range(func(k, v interface{}) bool {
+		pp, ok := v.(*PeerInfo)
+		if !ok {
+			utils.ErrorLogf("Invalid PP with network address %v in local PP list)", k)
+			return true
+		}
+
+		if pp.Status == PEER_UNREACHABLE {
+			return true
+		}
+
+		line := []string{
```

Comment:
instead of csv file, maybe json will make it easier for future usage when we need to provide this pplist to front-end. 
but don't need to change now, please just add a //todo here